horovod/common/ops/nccl_operations.cc

```diff
@@ -43,6 +43,13 @@ void NCCLContext::ErrorCheck(std::string op_name, ncclResult_t nccl_result) {
   }
 }
 
+void NCCLContext::ShutDown(){
```

Comment:
@Sergei-Lebedev, can you add `cudaStreamSynchronize()` before shutting down NCCL communicators?

The reason for this is that there could be outstanding work enqueued to NCCL, and I'm not sure if it's safe to destroy NCCL communicators while there's enqueued work.

After you make this change, can you check if `init.cu:645 WARN Cuda failure 'driver shutting down'` appears?

Reply:
@alsrgv, I think there is no need to explicitly wait for operations to be completed as long as NCCL do it internally. According to nccl documentation ncclCommDestroy waits for any uncompleted operations before destroying communicator. 
https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/comms.html#ncclcommdestroy

Reply:
Good find!  In this case, can you remove the comment in the operations.cc?